Construct T = Finv_Fprime * Fz

jburba opened the pull request:
This pull request reworks the matrix construction stack to construct the matrix product `Finv_Fprime = Finv * Fprime` in place instead of building `Finv` and `Fprime` independently.  This saves a lot of RAM usage as `Finv_Fprime` is constructed incrementally from the dense sub-matrices that comprise `Finv` and `Fprime`.  This also is also time efficient because doing dense matrix multiplications allows us to take advantage of the built-in threading within `numpy`.  For example, with an EoR FoV=12.9 deg (`nu=nv=15`) and a foreground FoV=30 deg (`nu_fg=nv_fg=29`), it takes 1.75 hours to compute `T = Finv * Fprime_Fz` because many of the matrix multiplications use the `scipy.sparse` library which does not allowing threading.  With 8 threads, this same matrix stack computed as `T = Finv_Fprime * Fz` (this pull request), the matrix compute time was 7 minutes.

# Testing

## Matrix comparisons

As a first test, I computed `Finv` and `Fprime` on the main branch and computed `Finv_Fprime_main = Finv * Fprime` and compared this with the `Finv_Fprime_branch` matrix as built in the `Finv_Fprime` branch.  For this test, I used matrices with matching EoR and foreground model FoV values of 12.9 deg as the matrices build quickly.  I then generated a model (u, v, f) vector and multiplied this vector by `Finv_Fprime_main` and `Finv_Fprime` and compared the resulting vectors.

<details>
<summary>The code used to generate this model (u, v, f) vector and the resulting plot is available here for posterity.</summary>

```
np.random.seed(912378)

# Construct a rectilinear (l, m, f) cube which can then be FFTd
# to the EoR model uv-plane and then vectorized to get the right
# Hermitian symmetry
lmf_cube_eor = np.random.normal(0, 1, (nf, nv, nu))

axes = (1, 2)
uvf_cube_eor = np.fft.ifftshift(lmf_cube_eor, axes=axes)
uvf_cube_eor = np.fft.fftn(uvf_cube_eor, axes=axes)
uvf_cube_eor = np.fft.fftshift(uvf_cube_eor, axes=axes)
# Remove the (u, v) = (0, 0) component from the EoR uv-plane
uv00_mask = np.zeros_like(uvf_cube_eor, dtype=bool)
uv00_mask[:, nv//2, nu//2] = True
uvf_vec_eor = uvf_cube_eor[~uv00_mask].flatten()
print(f"{uvf_vec_eor.shape      = }")

# Construct a rectilinear (l, m, f) cube which can then be FFTd
# to the FG model uv-plane and then vectorized to get the right
# Hermitian symmetry
lmf_cube_fg = np.random.normal(1e3, 100, (nf, nv_fg, nu_fg))

axes = (1, 2)
uvf_cube_fg = np.fft.ifftshift(lmf_cube_fg, axes=axes)
uvf_cube_fg = np.fft.fftn(uvf_cube_fg, axes=axes)
uvf_cube_fg = np.fft.fftshift(uvf_cube_fg, axes=axes)

fg_uv_mask = np.ones_like(uvf_cube_fg, dtype=bool)
if not ffm:
    fg_uv_mask[:, nv_fg//2, nu_fg//2] = False
uvf_vec_fg = uvf_cube_fg[fg_uv_mask].flatten()
print(f"{uvf_vec_fg.shape       = }")

# Concatenate the EoR and FG vectors
uvf_vec_sum = np.concatenate((uvf_vec_eor, uvf_vec_fg))
print(f"{uvf_vec_sum.shape      = }")
print(f"{Fprime_main.shape[1:]  = }")

# Alternatively, use a white noise vector
# np.random.seed(912673)
# mean = 1e-6
# var = 1.003333e1
# x = (
#     np.random.normal(mean, var, Fprime.shape[1])
#     + 1j*np.random.normal(mean, var, Fprime.shape[1])
# )

x = uvf_vec_sum

Finv_Fprime_x_main = Finv_Fprime_main * x
Finv_Fprime_x_branch = Finv_Fprime_branch * x

fig, axs = plt.subplots(2, 2, figsize=(8, 6), gridspec_kw={'height_ratios': [2, 1]})

funcs = [np.real, np.imag]
func_lbls = ['Real', 'Imag']

for (axs_col, func, func_lbl) in zip(axs.T, funcs, func_lbls):
    ax = axs_col[0]
    ax.set_title(func_lbl)
    ax.set_ylabel(r"$\mathbf{F}^{-1}\mathbf{F}'\boldsymbol{x}$")
    ax.plot(func(Finv_Fprime_x_main), alpha=0.5, label='Main')
    ax.plot(func(Finv_Fprime_x_branch), alpha=0.5, label='Finv_Fprime')
    ax.legend()

    ax = axs_col[1]
    ax.set_ylabel('Main - Finv_Fprime')
    ax.plot(func(Finv_Fprime_x_main) - func(Finv_Fprime_x_branch), 'k-')

fig.tight_layout()
```
</details>

This generated the following output figure.  The left and right columns show the real and imaginary components of the matrix-vector product `Finv_Fprime * x` when using `Finv_Fprime_main` (blue) and `Finv_Fprime_branch` (orange).  The bottom subplots show the residuals `Finv_Fprime_main*x - Finv_Fprime_branch*x`.  The results between the two branches agree well suggesting that the `build_Finv_Fprime` branch is working properly.

![image](https://github.com/user-attachments/assets/c0802471-01f4-436c-bbc4-794bb911052b)


## Full analyses

The plots below compare full BayesEoR analyses comparing the results run for identical input data and model parameters for two different foreground FoV values (12.9 deg in the top plot and 30 deg in the bottom plot).  The mock EoR FoV is 12.9 deg in both cases.  In all cases, the FoVs of the EoR and foreground components are identical in both data and model.  The results obtained from the Finv_Fprime branch which compute `T = Finv_Fprime * Fz` using the new `build_Finv_Fprime` function agree in all k bins with the results from the main branch which computes `T = Finv * Fprime_Fz`.

![image](https://github.com/user-attachments/assets/3a93db2c-a589-429f-ae57-67225c864152)
![image](https://github.com/user-attachments/assets/1a2295b2-c760-47c8-b729-1d187e713f42)

# Changes

- Adds a new `build_Finv_Fprime` function which computes the matrix product `Finv * Fprime` from their dense sub-matrices without constructing `Finv` and `Fprime` independently
  - Fixes #49 
  - A step toward #53 
- Updates the matrix prerequisite and build methods dictionaries to construct `T` from `Finv_Fprime` and `Fz`
- Updates the `sampled_uv_vectors` function in `bayeseor.matrices.matrix_funcs` to allow for kwargs for the model uv-plane resolutions `du` and `dv`
- Adds a new `build_nudft_array` function in `bayeseor.matrices.matrix_funcs` to construct general non-uniform DFT matrices which is used in `build_Finv_Fprime`

I realize that it's unrelated to the above changes, but I also moved the driver script from the root directory `BayesEoR/` to `BayesEoR/scripts`.  If the driver script is run from the root directory of the repo, it will always pull first from the current working directory when importing `bayeseor`.  This is not always desirable, especially if a different version of `bayeseor` is already installed in a python environment.  Moving the driver script to the `scripts/` directory ensures that it uses the `bayeseor` version installed in the python environment.  This was a necessary change for me to test the matrix building which is why I added this change as part of this pull request.  Because this change is quite distinct from the purpose of this pull request, I'm happy to remove this change and reimplement this as a separate issue/pull request for better tracking.

jburba commented:
With these changes, I was able to successfully build the matrices for an all sky foreground model (FoV=180 deg, nu_fg=nv_fg=155) for the first time.  The full matrix stack took only 3:39:12 to build with 16 threads (CPU cores), which is a massive speedup that's comparable to, if not better than, the speedup observed in #44.  For posterity, the MaxRSS reported by Slurm in this case was 373.06 GB.  Until now, attempts to build this all sky foreground model matrix stack were resulting in out of memory errors when requesting 1.5 TB of RAM.

PSims commented:
@jburba, this looks excellent!

After the following minor comments I think the PR is good to merge:
- Adding an equivalent of the the Finv_Fprime_x_main  vs Finv_Fprime_x_branch comparison as a unit test would be great (not necessary for this PR),  assuming a version with a small enough FoV can be defined that will run in a sensible testing interval.
- In `build_nudft_array`, a comment describing Nax and/or the last for loop would be useful. This could also just be an example (e.g. building multiple channels)
- For backwards compatibility (and possible future Dask functionality), it would be good to have a flag that, while defaulting to the combined build, allows building from `Finv` and `Fprime` if desired.

jburba commented:
I've made a few updates which I think satisfy your comments:

1. I've made a note of the unit test you proposed in my unit test branch todos.  That should be doable and is a good idea.  I'll add all of my notes on this to a new issue for adding unit testing in the repo.

2. I've added some examples to the docstring for `build_nudft_array`.  It should render nicely in the documentation.  If that's not what you had in mind I'm happy to move those examples elsewhere.  They also work as unit tests and can be added to test that function in the 1D and 2D Fourier transform case against `numpy.fft`.

3. I added a command line kwarg `--build-Finv-and-Fprime`, which despite being a bit lengthy, will toggle building `T = Finv * Fprime_Fz` (the legacy method) and write Finv and Fprime to disk.  If you exclude that flag (the default), it will calculate `T = Finv_Fprime * Fz` and only save the matrix product Finv_Fprime to disk.  I also tested that this kwarg works properly (which it does) and ran a quick check of T@x identical to the check I did in my original comment with Finv_Fprime@x.  The two different colored lines correspond to T constructed as `T = Finv * Fprime_Fz` (blue) and `T = Finv_Fprime * Fz` (orange).
   ![image](https://github.com/user-attachments/assets/c49f2cee-6bbf-4cf8-a196-54bbb52bcbd0)

Let me know what you think.  If all looks good, we can merge this into main.

PSims commented:
Looks great! Thanks Jacob, I think this is good to merge.
